### ISSUE TITLE
Improving Many-to-Many Reactor Communication

### ISSUE THREAD

**Issue by MattEWeber** [icyphy/lingua-franca#118]
I implemented the Chameneos benchmark from the Savina suite (see https://shamsimam.github.io/papers/2014-agere-savina.pdf) and the wiring got to be a little messy. See for yourself.

<img width="795" alt="Screen Shot 2020-03-12 at 8 53 36 PM" src="https://user-images.githubusercontent.com/3513451/76588196-a215b580-64a3-11ea-885b-3414ab652376.png">

In this model there are 5 lizard reactors (a Chameneos is a kind of lizard) receiving instructions from a mall reactor on how to exchange colors with each other. So the mall needs to be able to send and receive messages from each lizard, and each lizard needs to be able to send and receive messages with every other lizard. But on top of that, each lizard needs separate dedicated communication channels with the other lizards for both initiating and responding to color change requests. I'll show you why:

A two port ("send", "receive") implementation has a cycle.
<img width="635" alt="Screen Shot 2020-03-12 at 9 36 49 PM" src="https://user-images.githubusercontent.com/3513451/76590244-afce3980-64a9-11ea-9863-b3ed4e2e5c14.png">

A four port ("send", "receive", "respond", "ack") implementation doesn't.
<img width="528" alt="Screen Shot 2020-03-12 at 9 37 05 PM" src="https://user-images.githubusercontent.com/3513451/76590253-b361c080-64a9-11ea-8efe-d4090ebfd9eb.png">

Any ideas on how to make this easier to deal with? I'm thinking along the lines of higher order reactor patterns.

**Comment by lhstrh** [comment]
I'll look into these models tomorrow, but, yes, higher-order reactors is what we need. I have started discussions on this topic with George Ungureanu at KTH some time ago. He has done very cool work in this area and we should try to engage with him again. We're also in a much better position to do that now because the compiler and runtime infrastructure has gotten more mature.

**Comment by edwardalee** [comment]
One idea might be to resurrect the concept of realizing something like a remote procedure call.
The syntax could look something like this (consider this a very rough draft):
```
     reactor Lizard {
          input mall:int;
          io r:int -> int; // Functions as both input and output.
          reaction(mall) -> r {=
               set(r, mall);
          =} response(r) {=
               printf("Response is %d.\n", r);
          =}
          reaction(r) {=
              set(r, r+1);
          =}
    }
    main reactor Chameneos {
        m = new Mall();
        liz1 = new Lizard();
        liz2 = new Lizard();
        m.out -> liz1.mall;
        m.out -> liz2.mall;
        liz1.r <-> liz2.r;
    }
```
The above could be desugared as your design, I think.
A simpler use of the same mechanism would be a "service" as in SoA:
```
    reactor Service {
         io s:int->int;
         reaction(s) {=
              set(s, s + 1);
         =}
    }
```
This just responds to requests and never initiates them.
A user of the service would look like:
```
    reactor User {
         io u:int->int;
         reaction(startup) -> u {=
              set(u, 42);
         =} response(u) {=
               printf("Response should be 43: %d.\n", u);
         =}
    }
    main reactor Foo {
         u = new User();
         s = new Service();
         u.u <-> s.s;
    }
```

**Comment by mew2ub** [comment]
I think the biggest thing that would simplify this would be some kind a shorthand notation for creating and connecting iterated components. Each of the five lizard instances in this example is basically identical. Each lizard is connected to each of its neighbors, and each lizard should respond similarly if it is instructed to communicate with lizard "x" (unless this is lizard "x"). If there was some kind of notation to symbolically refer to a port as "a communication channel to lizard x" and let the code generator figure out how to handle the wires, that would be great.

I like this io port idea because it hides some of the complexity. To make sure I understand how this would work ... When you reference the port as an output you're referring to it as the "send" port and when you reference the port as an input you're referring to the "response" port, right? I think the example is missing "receive" and "ack" ports and should be modified to this:

```
     reactor Lizard {
          input mall:int;
          io r:int -> int; // Input and output for initiating communication.
          io s:int -> int; // Input and output for responding to communication.
          reaction(mall) -> r {=
               set(r, mall);
          =} response(r) {=
               printf("Response is %d.\n", r);
          =}
          reaction(r) {=
              set(r, r+1);
          =}
          reaction(s) {=
              set(s, s+1);
          =}
    }
    main reactor Chameneos {
        m = new Mall();
        liz1 = new Lizard();
        liz2 = new Lizard();
        m.out -> liz1.mall;
        m.out -> liz2.mall;
        liz1.r <-> liz2.s;
        liz2.r <-> liz1.s;
    }
```

Now the io port `r` corresponds to both "send" and "response" and the io port `s` corresponds to both "receive" and "ack".

**Comment by lhstrh** [comment]
At the time of writing the CyPhy paper, Inigo and I worked on a proof that the reactor model, by leveraging its dependency analysis, can guarantee deadlock-free remote procedure calls. We didn't report our findings in the CyPhy paper, but I do plan to incorporate it in my thesis.

**Comment by edwardalee** [comment]
Perhaps it would be better to make a syntactic distinction between a service *provider* and a service *invoker*. In your example, Lizard is both a provider and an invoker. Like this:
```
    reactor Lizard {
          input mall:int;
          service s:int -> int;
          invoke r:int -> int;
          reaction(mall) -> r {=
               set(r, mall);
          =} response(r) {=
               printf("Response is %d.\n", r);
          =}
          reaction(s) {=
              return(s, s+1);
          =}
    }
    main reactor Chameneos {
        m = new Mall();
        liz1 = new Lizard();
        liz2 = new Lizard();
        m.out -> liz1.mall;
        m.out -> liz2.mall;
        liz1.r <-> liz2.s;
        liz2.r <-> liz1.s;
    }
```
The return value of the reaction to `s` would have to be intercepted by the framework and sent back to the service invoker.

**Comment by cmnrd** [comment]
I like the idea of defining services. This would also fit well to the AUTOSAR line of work that we started with the DATE paper. If this would be added to LF, I think it might even make sense to go one step further. In your proposed code, a service is a single method. I think it would be vary helpful for more complex programs to group functionality (i.e. multiple methods) in a single service interface. This would even further reduce the connections that need to be established.

**Comment by mew2ub** [comment]
Here's the sketch of higher order syntax for services I described in the call today with @lhstrh, @akella84, and @edwardalee. This is all intended as syntactic sugar for simplifying connections between reactors. I realize the sketch doesn't stand on its own, so if you weren't in the meeting and have any questions, shoot me a message.

I'm going to hold off on doing more work on this until Marten shares the work on deadlock free procedure calls. 
```
------------------------------------------------------------
BASIC I/O SERVICE:

reactor Service {
    service s:int->int; // Have to declare type of input and output
    reaction(s) -> s {=
        set(s, s + 1);
    =}
}

reactor Client {
    timer t;
    invoke r:int->int; // Perhaps change name, see UML terminology
    reaction(t) -> r {
        set(r, 1);
    }
    reaction(r) {
        print(r);
    }
}

main reactor Container {
    a = new Service();
    x = new Client();

    x.r <-> a.s // or a.s <-> x.r?
}

------------------------------------------------------------
MULTIPLE CLIENTS CAN USE THE "SAME" SERVICE (different instances of ports and reactions)

main reactor Container {
    a = new Service();
    x = new Client();
    y = new Client()

    x.r <-> a.s
    y.r <-> a.s

    // Priority for service reactions is set in the order of connections?
    // Probably have to label a service's reaction with a keyword if it should be duplicated for multiple instances

}

------------------------------------------------------------
SERVICE CHAIN

reactor Service1 {
    service s:int->int;
    reaction(s) {=
        set(s, s + 1);
    =}
}

reactor Service2 {
    service s:int->int;
    reaction(s) -> s {=
        set(s, s + 2);
    =}
}

reactor Client {
    timer t;
    invoke r: int->int;
    reaction(t) -> r {
        set(r, 1);
    }
    reaction(r) {
        print(r);
    }
}

main reactor Container {
    a = new Service1();
    b = new Service2();
    x = new Client();

    x.r <-> (a.s -> b.s) 
    // Connects x to a to b to x.
}

------------------------------------------------------------
SERVICE RECURSION?

can you connect a service to itself with an after 0?
I don't really see the use case...


------------------------------------------------------------
ARRAYS OF CLIENTS

main reactor Container {
    a = new Service();
    for (i = 0; i < 20; i++) {
        clientArray[i] = new Client(parameter = i);
        clientArray[i].r <-> a.s;
    }
}
// We may want to simplify the for loop to avoid adding a math expression language.

```

**Comment by lhstrh** [comment]
This is an interesting proposal.

What is described above as a "basic I/O service" looks a bit more like a coroutine (and less like a remote procedure call). The end of the first reaction demarcates what in a coroutine would be a "yield." The key difference is that in coroutines, the yielding process _suspends_, and it resumes when the other routine yields. Here, however, the yielding reactor would loose any state that was sitting on the stack during the first reaction --- unless it explicitly gets saved using state variables. It is this cumbersome marshaling of data between reaction-scope variables and reactor state that we had figured _synchronous calls_ between reactors (that can occur anywhere inside of a reaction) could help avoid. It is a common scenario for one reactor to need a quick exchange with another before it can carry on with its reaction. In fact, the Chameneos benchmark seems to fit this description. This is what Inigo and I had been working on. Note that such synchronous calls would amount to a modest extension of the model (not just syntax). In the actor world, where _everything_ is asynchronous, folks use futures/promises to get around this issue.

The multiple clients that invoke the same "service" pose a challenge in the reactor model, because, at any given logical time, any distinct reaction may only execute once, and input can only be connected to at most one upstream port. As Matt suggests, a solution would be to create ports and reaction instances on demand (i.e., in response to a connection being made to them). Implementing this statically (as an AST transformation) would be fairly straight-forward, but the implications of allowing such transformations at runtime (i.e., during a mutation) would need to be carefully studied. This is definitely something worth exploring, however, and would be particularly relevant for the AUTOSAR work with Christian. That said, a native transformation, where reaction instances would be replicated for each connection, would pose a lot of unnecessary overhead. Instead, we should probably find a way to reuse the same instance while maintaining the property that each distinct reaction only executes once.

I am guessing that services would have to be constrained to only receive input from a single (two-way) port. A connection could either be routed back to its origin (client/server, `<->`) or routed to a subsequent reactor (`->`). However, I don't quite understand the "service chain" example. The connection `x.r <-> (a.s -> b.s) ` connects the port `x.r` to `(a.s -> b.s)`, which is not a port
but a connection. If I were to interpret `(a.s -> b.s)` as a function `(Port x Port) => Port` (to make it type check) then what would its implementation be?

It also occurs to me that while the decoupling of request sender (first reaction; before yield) and response handler (second reaction; after yield) it is _possible_ to get a synchronous response, but it is not required to happen, which is an interesting difference with the synchronous calls. The also begs the question: what would the semantics of `a.y <-> b.x after 25 msec` be? The logical delay could either be added to the call, to the response, or be split between both. But whereas `a.y -> b.x after d` means that if an `a.y` output was produced at `t`, the logical time at which reactions sensitive to `b.x` are to occur at `t + d`, there is no such guarantee in the `<->` variant. Specifically, in the `<->` case, there is no guarantee that if an `a.y` output is produced at `t` there will be an input observed at `a.y` at `t + d`. I am not sure how to reconcile this difference. For synchronous calls, on the other hand, the answer is simple: delays are not possible.

On the topic of higher-order reactors, it may be more helpful to look at _templates_ rather than pursue adding imperative-style code in LF for building patterns. I have been having discussions with George Ungureanu at KTH about this. Section II of this paper "Formal Design, Co-Simulation and Validation of a Radar Signal Processing System" provides some background on this idea of using
templates. After the last conversation I had with George a week or two ago, we agreed that it was probably easiest to experiment by first building some prototypes directly in TypeScript. It will allow us to quickly gain experience without having to tinker with the compiler, which is more slowly to develop. I took this as an action item, but help is welcome. My suggestion is to first run some trials (the benchmarks are a great start) before we would start translating some of these things from run-time to compile-time and come up with a suitable syntax. My plan is to do the same with the synchronous calls. I hope to have a working implementation of the synchronous calls ready for next weeks' group lunch...

**Comment by edwardalee** [comment]
One possible answer to the ‘after’ question is this:

    a.y <-> b.x after 10 msec, 10 msec;

**Comment by cmnrd** [comment]
Sorry about my long message. I have some thoughts on this issue and similar to Marten before, I think it is good to have these thoughts written down. I am happy to discuss on it here or in our next meeting.

In our discussion on Friday, there was a consensus that many-to-many communication requires duplication of input ports and reactions in service reactors. This brings the problem of replicating structure including ports and reaction bodies as well as blowing up the dependency DAC. This could have a severe impact on performance and scalability. We also discussed that this mechanism would add an implicit dependency between any two clients using the same service. I am not sure anymore why exactly this is the case, but having such implicit dependencies challenges composability. While thinking about this problem, I came up with an alternative solution that does not require this kind of replication.

Consider this example (syntax as was proposed above):
```
Cpp;
reactor Service {
  service s:int -> int;
  reaction(s) -> s {=
    s.set(*s.get() + 42);
  }=
}
reactor Client(x:int(42)) {
  invoke r:int -> int;
  reaction(startup) {=
    r.set(42);
  =}
  reaction(r) {=
    std::cout << x << " + 42 = " << *r.get() << '\n';
  =}  
}
main reactor Main {=
  s = new Service();
  c1 = new Client(x=1);
  c2 = new Client(x=2);
  c3 = new Client(x=3);
  // all three clients use the same service
  c1.r <-> s.s;
  c2.r <-> s.s;
  c3.r <-> s.s;
=}
```

How could this be realized using the syntax we currently have? There is a mechanism that allows us to build the service without replicating the service's reaction.  We can us a reaction to multiple inputs like this:
```
Cpp;
reactor Service {
  input s_in1:int;
  input s_in2:int;
  input s_in3:int;
  output s_out1:int;
  output s_out2:int;
  output s_out3:int;
  reaction(s_in1, s_in2, s_in_3) -> s_out1, s_out2, s_out3 {=
    if(s_in1.is_present()) {
      s_out1.set(*s_in1.get() + 42);
    }
    if(s_in2.is_present()) {
      s_out2.set(*s_in2.get() + 42);
    }
    if(s_in3.is_present()) {
      s_out3.set(*s_in3.get() + 42);
    }
  }=
}
reactor Client(x:int(42)) {
  output r_out:int;
  input r_in:int;
  reaction(startup) {=
    r_out.set(42);
  =}
  reaction(r_in) {=
    std::cout << x << " + 42 = " << *r_in.get() << '\n';
  =}  
}
main reactor Main {=
  s = new Service();
  c1 = new Client(x=1);
  c2 = new Client(x=2);
  c3 = new Client(x=3);
  // all three clients use the same service
  c1.r_out -> s.s_in1;
  c2.r_out -> s.s_in2;
  c3.r_out -> s.s_in3;
  s.s_out1 -> c1.r_in
  s.s_out2 -> c2.r_in
  s.s_out3 -> c3.r_in
=}
```
This effectively avoids the replication of reactions. No matter how many connections a service has, there is only one reaction. However, input ports are still replicated and there is code duplication in the if structure of the service's reaction bodies. We can solve this by introducing a new concept: the *multi-input port*. By multi-input port I mean a single port instance that represents multiple reactor ports and handles multiple connections simultaneously. This would be like an array of ports, where the array is extended for each additional connection to the service. Calling 'get()' on a multi-input port can yield multiple values if multiple clients send a value at the same logical time. 'get()' could for instance return a list of all values that are present at the current logical time. With such a port, the service can be implemented like this:
```
reactor Service {
  service s:int -> int;
  reaction(s) -> s {=
    for(int x : *s.get()) { // s.get() returns a pointer to a list of integers
      s.set(x + 42);
    }
  }=
}
```

Note that there is a problem with the code above. It is entirely unclear to which client port the service's response should be routed to. When replicating both ports and reactions, it is always clear to which port a response needs to be send. This is not the case here as this information is not part of the reactor structure anymore. But we can encode this information in the data (which in my opinion is a better approach than hardcoding it in the program structure). Instead of a plain `int`, the service could yield a request object like `request<int>`. This could be as simple as a pair of the actual value and an ID representing the sending client. Further, there could be a method `make_response` that takes the initial request and a value and produces a response object. With this the service could be implemented like this:
```
reactor Service {
  service s:int -> int;
  reaction(s) -> s {=
    for(request<T> req : *s.get()) { // s.get() returns a pointer to a list of request objects
      s.set(req.make_response(req.value + 42);
    }
  }=
}
```

With this solution, a service with multiple clients can be realized using a single port instance and a single reaction. Only the data is replicated, not the program structure that handles it. As far as I can see, this approach would also prevent the introduction of unexpected dependencies between the clients. Furthermore, I believe that the multi-port concept would make mutations quite simple. It would be a matter of dynamically adding or removing a port from its internal port list.

A note about determinism and preserving reactor semantics: This probably requires a deeper thought, but I think the multi-input port would preserve the reactor semantics. After all it is just a shortcut for handling multiple input ports. The multi-input port is a collection of plain reactor input ports (which I will call *logical* ports) that each have a single source. If at least one of the logical inputs is present, the corresponding reaction is triggered and has access to all of the logical inputs present at the given logical time. This is the same semantics as in my second code snippet above. In order to maintain determinism, we further need to establish an order in which values are retrieved from the logical ports contained within the multi-input port. This can be any well-defined order. For instance it could be the order in which connection where established. The ordering problem also exists in the case where reactions are replicated, but there the order is inferred from the dependency DAC.

**Comment by lhstrh** [comment]
Arguably, the replication scheme also establishes something that is essentially a multiport. However, exposing the multiplexed structure so that it can be iterated over _within_ a service reaction rather than letting the runtime iteratively executing the same reaction, each time with different input/output, is an interesting idea. It does simplify the dependency graph (because the dependencies between different invocations are no longer reflected in it), although there will still be a dependency from the service reaction to each upstream reaction connected to it.

A for loop will probably be more efficient than a multitude of reaction invocations, but I suspect that the synchronization burden of having ordered service requests would typically dominate the execution time associated with the service reaction invocation(s). And whereas the for loop imposes a barrier synchronization among _all_ clients, this is avoided in the replication scheme -- allowing for more parallelism. There is also an advantage to this idea of writing a service as if it served a single client, and then letting the compiler to "lift" this implementation into a one that serves many. This shifts some complexity from the programmer to the compiler (and/or runtime). 

I think you're right to point out that dependencies pose a challenge to scalability. But without them, we loose determinism. To address this tension, and to acknowledge there are many scenarios in which determinism is not required, we could introduce syntax (maybe `<~>`) that denotes an explicit
relaxation of the imposed ordering constraints between service invocations. The philosophy, however, should remain that determinism is the default, and it can be relinquished explicitly for performance considerations. Note that with the for-loop solution we actually don't have any way to enforce an order (it is controlled by the programmer), meaning also that we cannot relax it without changing the reaction implementation. I think this is a significant downside of the for-loop approach.

That said, while the replication scheme ensures us that all the necessary dependencies are in place to guarantee determinism, it does so in a rather brute-force and inefficient way. As I mentioned earlier, operationally, there is no real need to create separate reaction instances for each client/server pair. We could implement optimizations that are preserving of the reactor semantics.

**Comment by cmnrd** [comment]
> And whereas the for loop imposes a barrier synchronization among all clients, this is avoided in the replication scheme -- allowing for more parallelism.

True. I didn't think of that. But, I don't think the two strategies are exclusive. With the multi-input port approach we could make the for loop implicit and write:
```
reactor Service {
  service s:int -> int;
  reaction(s) -> s {=
    request<T> req = *s.get()) // s.get() returns a pointer to a request object
    s.set(req.make_response(req.value + 42);
  }=
}
```
This would trigger the reaction to 's' once for each request received. Now this is very similar to the replication approach. In fact the reaction needs to be replicated since it can only be triggered once at the same logical time. However, I argue that this replication could happen implicitly at runtime. The reaction could be a *multi-reaction* that appears once on the dependency DAC but in fact represents multiple reaction instances. In other words, the multi-reaction is a reaction that can be triggered multiple times following some well-defined rules. This approach would do the replication as needed at runtime, but not in the structure of the reactor program itself. To me it seems as if this combines the best of the two strategies discussed above.

In fact, the multi-input port could provide multiple strategies for triggering the reaction. It could trigger them in order as soon as a request arrives (this would be similar to the replication strategy), or with a relaxed order and possibly parallel execution of the different reaction instances. This would also move the selection of a execution strategy to a single place. With the syntax you suggested (`<~>`) the strategy would be selected per connection. But to me it is unclear what would happen if one client is connected using `<->` and another client using `<~>`.

**Comment by mew2ub** [comment]
Maybe we should be considering both the multi-reaction idea (one reaction per connection) and the multi-port idea (one reaction variable per connection) as different flavors of services. So far it has been discussed as a weakness, but the synchronization aspect of multi-port services makes it possible to look at all the ports at once. Consider a service that returned to each client the total number of its current clients. You couldn't implement that as a multi-reaction. Alternatively, multi-reactions would be more efficient for implementing side-effect free services that can react at any time.

I imagine map-reduce is going to be one of the more valuable patterns for services, and it provides a scenario for both styles: The map operation could be defined as an independent multi-reaction service, and the reduce operation could be a synchronized multi-port service.

**Comment by lhstrh** [comment]
A quick update: while I don't see a way of effectively using _synchronous calls_ to exchange data in a many-to-many scenario like the Chameneos benchmark (it would lead to cycles in the dependency graph), for the quick sidestep to gain access to a neighbor's state, it works rather well. In the PingPong benchmark, for instance, it results in roughly an order of magnitude reduction in execution time (and a significant reduction in code size).

**Comment by cmnrd** [comment]
I took a run at the dining philosophers benchmark from the Savina suite and added 3 different C++ implementations. It turned out that the reactor implementation has some subtle trade-offs and I want to discuss them here.

Lets have a look at a philosopher first:
```
reactor Philosopher(id:unsigned(0), count:unsigned(10000), 
                    starvation_threshold:unsigned(1000)) {
    input eat:void;
    input denied:void;
    output hungry:void;
    output done:void;
    
    state times_eaten:unsigned;
    state times_denied:unsigned;
 
    logical action start_eating;
    
    reaction(startup) -> start_eating {=
        std::cout << "Hello! I am philosopher " << id << " and I am very hungry!\n";
        times_eaten = 0;
        times_denied = 0;
        // start eating right away
        start_eating.schedule();   
    =}
    
    reaction(start_eating) -> hungry {=
        // signal that I am hungry
        hungry.set();
    =}
    
    reaction(eat) -> done, start_eating, finished {=
        // ... take left and right fork
        std::cout << "Philosopher " << id << " is eating.\n";
        times_eaten++;
        times_denied = 0;
        // ... put forks away
        done.set(); // signal that I am done eating
        std::cout << "Philosopher " << id << " is thinking.\n";

        if (times_eaten < count) {
            // Actually, I feel hungry again...
            start_eating.schedule();
        } 
    =}
    
    reaction(denied) -> start_eating {=
        times_denied++;
        if (times_denied == starvation_threshold) {
            std::cout << "Philosopher " << id << " is starving!\n";
        }

        // Well, I will just try again...
        start_eating.schedule();
    =}
}
```
I simplified the code above a bit. The actual benchmark cose is a bit more complex,
but those additional features are not relevant here. The philosophers in the Savina benchmark are always hungry. Each iteration (triggereing of the `start_eating` action), they send a `hungry` request to the arbitrator. The arbitrator replies ither with `eat` granting permission to take the forks and eat, or with `denied` telling the philosopher that his forks are currently not available. When the philosopher finishes eating, it notifies the arbitrator via the `done` port. In this reactor implementation all philosophers are synchronous. This means in each logical time step, all reactors send the `hungry` signal.

So far so good. Looking at the Arbitrator, things get a bit more interesting. One problem is quite obvious: We need a mechanism to instantiate and connect multiple reactors at once. Doing this by hand for 20 philosophers is pain.

In my first try ([PhilosophersSequential.lf](https://github.com/icyphy/lingua-franca/blob/a7c5ab488fac7560ca6726aac0b85e8d90275f4c/benchmark/Cpp/Savina/PhilosophersSequential.lf)) I did what we called the 'replication approach'. For each `hungry` port of the philosophers there is one reaction that checks if the forks are available and replies accordingly with `eat` or `denied`. Similar, there is a reaction for each `done` port of the philosophers that frees the previously allocated forks. I did it such that first all the `hungry` reactions for all philosopher appear in the program and then all `done` reactions for all philosophers. I think this is also how the automatic replication would do it (BTW: placing the `done` reactions before the `hungry` reactions would create a dependency cycle). It looks like this:
```
    reaction(p0.hungry) -> p0.eat, p0.denied {=
        if (acquire_forks(forks, 0)) {
            p0.eat.set();
        } else {
            retries++;
            p0.denied.set();
        }
    =}
    \\ similiar reactions for the other 19 philosophers

    reaction(p0.done) {=
        free_forks(forks, 0);
    =}
    \\ similiar reactions for the other 19 philosophers
```
But there is a problem with this ordering of reactions. First, `hungry` from philosopher 0 will be processed and it will get access to the forks. Then, `hungry` from philosopher 1 is processed, and its request will be denied as philosopher 0 already uses one fork. Philosopher 2 will then get access and philosopher 3 will be denied, and so on... . The `done` requests will only be processed after all the`hungry` requests have been processed. Due to the deterministic ordering, this means that even philosophers alway get the forks and oddd philosophers starve until the even philosopher finished eating.

[PhilosophersSequentialInterleaved.lf](https://github.com/icyphy/lingua-franca/blob/a7c5ab488fac7560ca6726aac0b85e8d90275f4c/benchmark/Cpp/Savina/PhilosophersSequentialInterleaved.lf) fixes this problem by reordering the reactions. It interleaves the `hungry` and `done` reactions. The highest priority has `hungry` from philosopher 0, then comes `done` from philosopher 0, then `hungry` from philosopher 1, then `done` from philosopher 1 and so on. This effectively reorders the processing of incomming messages such that all the `hungry` requests of the philosophers can be granted. This ordering ensures that `done` from philosopher N is processed before `hungry` from philosopher N+1 and, therefore, it is guaranteed that the resources used by philosopher N are freed by the time the next `hungry` request is processed. This implementation of the benchmark is very efficient as it does not send any `denied` messages at all.

So, all good? The above only works well for this simple benchmark where the philosophers do
not perform any work. The implementation above is strictly sequential. Eating is mutual exclusive
and the philosophers take turns in round robin manner. This certainly is a valid solution to the problem, but it is very inefficient if eating takes a long time. How can we allow multiple philosophers to eat in parallel? I solved this in [PhilosophersSingleReaction.lf](https://github.com/icyphy/lingua-franca/blob/a7c5ab488fac7560ca6726aac0b85e8d90275f4c/benchmark/Cpp/Savina/PhilosophersSingleReaction.lf). This implementation uses a single reaction that is triggered by all `hungry` ports and sends replies to all philosophers. I used a C++ hack to be able to iterate over all philosophers in a for loop. Seeing all `hungry` requests in a single reaction, allows the arbitrator to find a good solution that lets as many philosophers eat as is possible (Since in this benchmarks the philosophers are always hungry, the strategy becomes trivial). 

So what do we learn from this?
1. At least for some problems it is beneficial to see all incoming requests  in one single reaction rather than having independent reactions for each request.
2. Reaction replication might have unexpected outcomes, because it is difficult to control the
way the generated reaction instances are ordered.
3. The determinism helps in the problem solution. It reduces the number of denied messages by some orders of magnitude in contrast to the actor implementations and even reduces it to zero in a sequential implementation.

**Comment by mryndzionek** [comment]
Interesting discussion going on here. I'm wondering how it all compares to the solution presented [here](https://www.state-machine.com/doc/AN_DPP.pdf)? In this solution there is no need for `denied` messages - philosophers just patiently wait for permission to eat. Philosophers can it in parallel, eating can take a long time and the serving order is preserved, so it's fair. When comparing/evaluating concurrency frameworks I'm always interested if [Cigarette smokers problem](https://en.wikipedia.org/wiki/Cigarette_smokers_problem) solution is possible and how simple it is. One interesting post I've found recently is [here](http://chemist.io/concurrency.html).

**Comment by cmnrd** [comment]
Closing this because with the multi ports we have a convenient solution.
